internal/spoke_k8s_client/factory.go

```diff
@@ -1,90 +1,255 @@
 type SpokeK8sClientFactory interface {
-	CreateFromSecret(secret *corev1.Secret) (SpokeK8sClient, error)
-	ClientAndSetFromSecret(secret *corev1.Secret) (SpokeK8sClient, *kubernetes.Clientset, error)
+	CreateFromSecret(ctx context.Context, secret *corev1.Secret) (SpokeK8sClient, error)
+	ClientAndSetFromSecret(ctx context.Context, secret *corev1.Secret) (SpokeK8sClient, *kubernetes.Clientset, error)
+}
+
+// SpokeK8sClientFactoryBuilder contains the data and logic needed to create a spoke client factory. Don't create
+// instances of this directly, use the NewFactory function instead.
+type SpokeK8sClientFactoryBuilder struct {
+	logger            logrus.FieldLogger
+	hubClient         ctrlclient.Client
+	transportWrappers []func(http.RoundTripper) http.RoundTripper
 }
 
 type spokeK8sClientFactory struct {
-	log logrus.FieldLogger
+	logger            logrus.FieldLogger
+	hubClient         ctrlclient.Client
+	transportWrappers []func(http.RoundTripper) http.RoundTripper
+}
+
+// NewFactory creates a builder that can then be used to configure and create a spoke client factory.
+func NewFactory() *SpokeK8sClientFactoryBuilder {
+	return &SpokeK8sClientFactoryBuilder{}
+}
+
+// SetLogger sets the object that will be used to write to the log. This is mandatory.
+func (b *SpokeK8sClientFactoryBuilder) SetLogger(value logrus.FieldLogger) *SpokeK8sClientFactoryBuilder {
+	b.logger = value
+	return b
+}
+
+// SetHubClient sets the client that will be used to call the API of the hub cluster. This is mandatory.
+func (b *SpokeK8sClientFactoryBuilder) SetHubClient(value ctrlclient.Client) *SpokeK8sClientFactoryBuilder {
+	b.hubClient = value
+	return b
+}
+
+// AddTransportWrapper adds a function that will be called to potentially modify the original HTTP transport. This is
+// optional and by default there are no wrappers.
+func (b *SpokeK8sClientFactoryBuilder) AddTransportWrapper(value func(http.RoundTripper) http.RoundTripper) *SpokeK8sClientFactoryBuilder {
+	b.transportWrappers = append(b.transportWrappers, value)
+	return b
 }
 
-func NewSpokeK8sClientFactory(log logrus.FieldLogger) SpokeK8sClientFactory {
-	return &spokeK8sClientFactory{
-		log: log,
+// Build uses the data stored in the builder to create and configure a new factory.
+func (b *SpokeK8sClientFactoryBuilder) Build() (result SpokeK8sClientFactory, err error) {
+	// Check parameters:
+	if b.logger == nil {
+		err = errors.New("logger is mandatory")
+		return
 	}
+	if b.hubClient == nil {
+		err = errors.New("hub client is mandatory")
+		return
+	}
+
+	// Create and populate the object:
+	result = &spokeK8sClientFactory{
+		logger:            b.logger,
+		hubClient:         b.hubClient,
+		transportWrappers: slices.Clone(b.transportWrappers),
+	}
+	return
+}
+
+func (f *spokeK8sClientFactory) CreateFromSecret(ctx context.Context, secret *corev1.Secret) (result SpokeK8sClient, err error) {
```

Comment:
I wonder why better to return result this way? 
(result SpokeK8sClient, err error) ?
I believe 99% of the code doesn't do it this way and i just wonder why it is better

Reply:
It is probably a matter of taste. I like to have the names of the return parameters: helps understand what to expect. Not very important in this case as the meaning is very clear. I can change it if you want.

Reply:
I agree that it is taste issue :) just most of the code have another style so why to have different styles?

Reply:
Fair enough, I will change it.

Reply:
Done.